students/GuyCChan/session09/objOriMailRoom.py

```diff
@@ -0,0 +1,392 @@
+import math
+import sys
+
+# don_db = DonorCollection(test_data())
+
+def test_data():
+    """ Data to be used as samples for coding and testing."""
+    return {"David Andrews" : [200.50, 400.00, 250.75],
+            "John Goodfellow" : [25.00, 175.50],
+            "Mary Suzuki" : [75.00, 125.00, 250.00],
+            "Bonney Lake" : [500.50, 700.75, 500.25],
+            "DeMarcus Rollins" : [155.00, 165.00]
+            }
+
+
+class Donor:
+    def __init__(self, name, donation=None):
+        """ Create Donor class for data and methods for individual donor."""
+        self.name = name.strip()
+        if donation is None:
+            self.donations = []
+        else:
+            self.donations = list(donations)
+
+    @staticmethod
+    def valid_donation(donation):
+        """ Make sure entered amount of a donation is valid before adding to database."""
+        if donation < 0:
+            raise ValueError("Donation has to be more than 0.")
+        if donation < 0.1:
+            raise ValueError("Donation has to be more than 10 cents")
+        return donation
+
+    def add_donation(self, amount):
+        amount = self.valid_donation(amount)
+        self.donations.append(amount)
+
+    @property
+    def num_donations(self):
+        return len(self.donations)
+
+    @property
+    def tot_donations(self):
+        return sum(self.donations)
+
+    @property
+    def avg_donation(self):
+        """ Average donation is total donations divided by the number of donations."""
+        return self.tot_donations / self.num_donations
+
+    @property
+    def last_donation(self):
+        try:
+            return self.donation[-1]
+        except IndexError:
+            return None
+
+    def thank_you_letter():
+        print("Dear {}, \n"
+            "Thank you for your generous gift of ${} to promote public education \n"
+            "and to support research about bats. \n"
+            "The Bat Society".format(self.name, self.last_donation))
+
+
+class DonorCollection:
+    def __init__(self, donors=None):
+        """ Initialize the DonorCollection class to hold info, data, and methods for all donors in database."""
+        if donors is None:
+            self.donor_inf = {}
+        else:
+            self.donor_inf = {name: donations}
+
+    @property
+    def donors(self):
+        return self.donors_inf.keys()
+
+    @property
+    def list_of_donors(self):
+        dlist = []
+        return self.donors.keys()
```

Comment:
this is going to return a "dict_keys" object, which can be iterated through, but is not a list. If you want a list:

return list(self. donors_inf)